regex-parsing/src/main/java/org/sonarsource/analyzer/commons/regex/finders/RedosFinder.java

```diff
@@ -0,0 +1,324 @@
+        return false;
+      }
+      RepetitionTree repetition = (RepetitionTree) state;
+      return repetition.getQuantifier().getMinimumRepetitions() == 0 && repetition.getQuantifier().isOpenEnded()
+        && canMatchAnyCharacter(repetition.getElement());
+    }
+
+    private boolean canMatchAnyCharacter(RegexTree tree) {
+      SimplifiedRegexCharacterClass characterClass = new SimplifiedRegexCharacterClass();
+      for (RegexTree singleCharacter : collectSingleCharacters(tree, new ArrayList<>())) {
+        if (singleCharacter.is(RegexTree.Kind.DOT)) {
+          characterClass.add((DotTree) singleCharacter);
+        } else {
+          characterClass.add((CharacterClassElementTree) singleCharacter);
+        }
+      }
+      return characterClass.matchesAnyCharacter();
+    }
+
+    private List<RegexTree> collectSingleCharacters(@Nullable RegexTree tree, List<RegexTree> accumulator) {
+      if (tree == null) {
+        return accumulator;
+      }
+      if (tree instanceof CharacterClassElementTree || tree.is(RegexTree.Kind.DOT)) {
+        accumulator.add(tree);
+      } else if (tree.is(RegexTree.Kind.DISJUNCTION)) {
+        for (RegexTree alternative : ((DisjunctionTree) tree).getAlternatives()) {
+          collectSingleCharacters(alternative, accumulator);
+        }
+      } else if (tree instanceof GroupTree) {
+        collectSingleCharacters(((GroupTree) tree).getElement(), accumulator);
+      } else if (tree.is(RegexTree.Kind.REPETITION)) {
+        RepetitionTree repetition = (RepetitionTree) tree;
+        if (repetition.getQuantifier().getMinimumRepetitions() <= 1) {
+          collectSingleCharacters(repetition.getElement(), accumulator);
+        }
+      }
+      return accumulator;
+    }
+
+  }
+
+  private void addBacktracking(BacktrackingType newBacktrackingType) {
+    if (newBacktrackingType.ordinal() < foundBacktrackingType.ordinal()) {
+      foundBacktrackingType = newBacktrackingType;
+    }
+  }
+
+  private class BacktrackingFinder extends RegexBaseVisitor {
+
+    private final boolean isReluctant;
+    private final AutomatonState endOfLoop;
+
+    public BacktrackingFinder(boolean isReluctant, AutomatonState endOfLoop) {
+      this.isReluctant = isReluctant;
+      this.endOfLoop = endOfLoop;
+    }
+
+    @Override
+    public void visitAtomicGroup(AtomicGroupTree tree) {
+      new RedosVisitor(tree, tree.continuation(), false, false).visit(tree);
+    }
+
+    @Override
+    public void visitRepetition(RepetitionTree tree) {
+      if (tree.isPossessive()) {
+        new RedosVisitor(tree, tree.continuation(), false, false).visit(tree);
+      } else if (containsIntersections(Arrays.asList(tree.getElement(), tree.continuation()))) {
+        BacktrackingType greedyComplexity = tree.getQuantifier().isOpenEnded() ? BacktrackingType.QUADRATIC_WHEN_OPTIMIZED : BacktrackingType.LINEAR_WHEN_OPTIMIZED;
+        addBacktracking(isReluctant ? BacktrackingType.ALWAYS_EXPONENTIAL : greedyComplexity);
```

Comment:
I think there is if-else if block will be easier to understand the logic:
```java
var greedyComplexity = BacktrackingType.LINEAR_WHEN_OPTIMIZED;
if (isReluctant) {
  greedyComplexity = BacktrackingType.ALWAYS_EXPONENTIAL;
} else if (tree.getQuantifier().isOpenEnded()) {
  greedyComplexity = BacktrackingType.QUADRATIC_WHEN_OPTIMIZED
}
```

Reply:
To be honest I find the original snippet actually easier to understand, so I'll leave it like that if it's okay with you.